### PR TITLE
update auto-property analyzer to explicitly check for exactly one setter statement

### LINKED ISSUE #5291: "CSharp.UseAutoProperty.UseAutoPropertyAnalyer" exceptions
These appeared in my error list so I don't have a dump:
```C#
Severity	Code	Description	Project	File	Line	Detail Description
Message	AD0001	Analyzer 'Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.	CodeAnalysis		1	Analyzer 'Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer' threw the following exception:
'System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer.GetGetterExpression(IMethodSymbol getMethod, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.UseAutoProperty.AbstractUseAutoPropertyAnalyzer`4.AnalyzeProperty(ConcurrentBag`1 analysisResults, SymbolAnalysisContext symbolContext)
   at Microsoft.CodeAnalysis.UseAutoProperty.AbstractUseAutoPropertyAnalyzer`4.<>c__DisplayClass13_0.<Initialize>b__1(SymbolAnalysisContext sac)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass33_0.<ExecuteSymbolActionsCore>b__0()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze)'.
```
```C3
Severity	Code	Description	Project	File	Line	Detail Description
Message	AD0001	Analyzer 'Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer' threw an exception of type 'System.InvalidOperationException' with message 'Sequence contains more than one element'.	CodeAnalysis		1	Analyzer 'Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer' threw the following exception:
'System.InvalidOperationException: Sequence contains more than one element
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty.UseAutoPropertyAnalyzer.GetGetterExpression(IMethodSymbol getMethod, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.UseAutoProperty.AbstractUseAutoPropertyAnalyzer`4.AnalyzeProperty(ConcurrentBag`1 analysisResults, SymbolAnalysisContext symbolContext)
   at Microsoft.CodeAnalysis.UseAutoProperty.AbstractUseAutoPropertyAnalyzer`4.<>c__DisplayClass13_0.<Initialize>b__1(SymbolAnalysisContext sac)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass33_0.<ExecuteSymbolActionsCore>b__0()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze)'.
```

### PR DESCRIPTION
**Pull request by brettfo**
Update UseAutoPropertyAnalyzers to handle a valid getter (e.g., one statement) but an invalid setter (e.g., 0 or 2+ statements).

**Examples:**
``` C#
int i;
int P
{
    get { return i; }
    set
    {
        int x = 1;
        i = value;
    }
}
```

``` VB
Dim i As Integer
Property P As Integer
    Get
        Return i
    End Get
    Set
        Dim x = 1
        i = value
    End Set
End Property
```

Fixes #5291.

### CODE REVIEW

**Comment by mmitche** [comment]
Can one of the admins approve this PR test or whitelist this user? ('@dotnet-bot add to whitelist' to whitelist, '@dotnet-bot okay to test' to accept for PR, '@dotnet-bot test this please' to retest.  Case sensitive).

**Comment by brettfo** [comment]
@Pilchie @DustinCampbell @dpoeschl @jasonmalinowski @rchande @balajikris @CyrusNajmabadi @jmarolf @davkean for review

File: src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs [hunk 1]
```diff
@@ -101,22 +101,22 @@ protected override ExpressionSyntax GetGetterExpression(IMethodSymbol getMethod,
...
-            //      set { field = value; } or
-            //      set { this.field = value; }
-            var firstStatement = setAccessor?.Body.Statements.SingleOrDefault();
-            if (firstStatement?.Kind() == SyntaxKind.ExpressionStatement)
+            var statements = setAccessor?.Body?.Statements;
+            if (statements?.Count == 1)
             {
-                var expressionStatement = (ExpressionStatementSyntax)firstStatement;
-                if (expressionStatement.Expression.Kind() == SyntaxKind.SimpleAssignmentExpression)
+                // this only works with a setter body with exactly one statement
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: You should probably keep the original comment, which was more descriptive about what pattern the code looks for.

**Comment by DustinCampbell** [comment]
:+1:

**Comment by brettfo** [comment]
Updated both C# and VB comments to be more descriptive about what patterns are acceptable.

File: src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs [hunk 2]
```diff
@@ -100,23 +103,26 @@ protected override ExpressionSyntax GetGetterExpression(IMethodSymbol getMethod,
...
-            if (firstStatement?.Kind() == SyntaxKind.ExpressionStatement)
+            //     set { field = value; } or
+            //     set { this.field = value; }
+            var setAccessor = setMethod.DeclaringSyntaxReferences[0].GetSyntax(cancellationToken) as AccessorDeclarationSyntax;
+            var statements = setAccessor?.Body?.Statements;
+            if (statements?.Count == 1)
             {
-                var expressionStatement = (ExpressionStatementSyntax)firstStatement;
-                if (expressionStatement.Expression.Kind() == SyntaxKind.SimpleAssignmentExpression)
+                var firstStatement = statements.Value[0];
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: just "statement", since there's only one? "first" is a bit confusing here.

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by Pilchie** [comment]
@dotnet-bot add to whitelist

**Pull request merged**